File: website/src/modules/api/generic.ts
import {UserRequest, UserResponse} from "./user";
import {ApiResponse, make_request, NewApiRequest} from "./api";

export type RequestData =
    { Ping: number }
    | { User: UserRequest };
export type ResponseData =
    ResponseSimple
    | { Pong: number }
    | { GenericError: PacketError }
    | { User: UserResponse };
enum ResponseSimple {
    GenericSuccess = "GenericSuccess",
}

export function check_error(response: ResponseData): PacketResult<void> {
    // @ts-ignore
    if ("GenericError" in response) {
        return response.GenericError;
    }
}
export type PacketResult<T> = PacketError | T;
export class PacketError {
    data: string;
    error_text: string;

    constructor(error_text: string) {
        this.data = JSON.stringify({client_error: new Error(error_text)});
        this.error_text = error_text;
    }
}

export async function ping(value: number): Promise<PacketResult<number>> {
    const response = await make_request(NewApiRequest({Ping: value}));
    const error = check_error(response.data);
    if (error) {
        return error;
    }
    // @ts-ignore
    if ("Pong" in response.data) {
        return response.data.Pong;
    } else {
        console.error(response);
        return new PacketError("Bad Packet");
    }
}
export async function make_user_request(request: UserRequest): Promise<PacketResult<ApiResponse>> {
    const response = await make_request(NewApiRequest({User: request}));
    const error = check_error(response.data);
    if (error) {
        return error;
    }
    return response;
}

File: src/api/generic.rs
use std::fmt::Debug;
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::api::token::AuthToken;
use crate::api::user::{handle_user_request, UserRequest, UserResponse};
use crate::database::connection_pool::ConnectionPool;

pub type Request = TokenPacket<RequestData>;
pub type Response = TokenPacket<ResponseData>;
impl<E> From<E> for Response where E: ReadableError {
    fn from(from: E) -> Self {
        Self {
            token: None,
            data: from.into(),
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TokenPacket<T> {
    pub token: Option<AuthToken>,
    pub data: T,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum RequestData {
    Ping(i64),
    User(UserRequest),
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum ResponseData {
    Pong(i64),
    GenericSuccess,
    GenericError {
        data: String,
        error_text: String,
    },
    User(UserResponse),
}
impl<E> From<E> for ResponseData where E: ReadableError {
    fn from(from: E) -> Self {
        let error_text = from.read();
        let data = match serde_json::to_string(&from) {
            Ok(data) => data,
            Err(error) => return Self::GenericError {
                data: "".to_string(),
                error_text: format!("Full failure, cause: {}, json error: {}", error_text, error),
            }
        };
        Self::GenericError { data, error_text }
    }
}

pub trait ReadableError: Debug + Serialize {
    fn read(&self) -> String;
}

pub async fn handle_request(request: Request, pool: Arc<ConnectionPool>) -> Response {
    match request.data {
        RequestData::Ping(value) => Response {
            token: None,
            data: ResponseData::Pong(value),
        },
        RequestData::User(user_request) => handle_user_request(request.token, user_request, pool).await,
    }
}
